### ISSUE TITLE
'replicate' padding in convolution is 77 times slower on cpu than 'zeros'

### ISSUE THREAD

**Issue by helion-du-mas-des-bourboux-thales** [pytorch/pytorch#68512]
## 🐛 Bug

When using a 2D convolution, 'replicate' padding is 77 times slower than 'zeros' padding on cpu.
I observe only a 30% difference on gpu. I would not have expected a so big difference on cpu.

```
import torch
import time

print(torch.__version__)

###
nb_trials = 10
kernel_size = 3
dilation = 2**8
nb_chs = 64
batch_size = 16
data_length = 28

devices = ['cpu','cuda:0']

dims = {
    '1d':{
        'tensor' : torch.rand(batch_size,nb_chs,data_length),
        'conv' : torch.nn.Conv1d,
    },
    '2d':{
        'tensor' : torch.rand(batch_size,nb_chs,data_length,data_length),
        'conv' : torch.nn.Conv2d,
    },
}

def get_time_perfs(nb_trials, verbose=True):
    for device in devices:
        for dim in dims.keys():

            a = dims[dim]['tensor']
            a = a.to(device)

            for pad in ['zeros', 'replicate']:

                conv = dims[dim]['conv'](nb_chs,nb_chs, kernel_size=kernel_size, padding_mode=pad,
                    padding=dilation*(kernel_size-1)//2, dilation=dilation)
                conv.to(device)

                time_start = time.time()
                for _ in range(nb_trials):
                    b = conv(a)
                time_end = time.time()

                if verbose:
                    print(f"{device}, {dim}, {pad} took {1000*(time_end-time_start)/nb_trials} ms")

    return

get_time_perfs(2, verbose=False)
get_time_perfs(nb_trials)
```

result:
```
1.10.0+cu102

cpu, 1d, zeros took 7.705163955688477 ms
cpu, 1d, replicate took 0.7227182388305664 ms

cpu, 2d, zeros took 6.098818778991699 ms
cpu, 2d, replicate took 472.0391035079956 ms

cuda:0, 1d, zeros took 10.245108604431152 ms
cuda:0, 1d, replicate took 0.14510154724121094 ms

cuda:0, 2d, zeros took 0.09889602661132812 ms
cuda:0, 2d, replicate took 0.13229846954345703 ms
```

**Comment by albanD** [comment]
There are a few things about your code:
- The cuda timings are all wrong because you don't properly `torch.cuda.synchronize()` to see how much time the computation actually took.
- A dilation of 2**8 is expected? That is a lot bigger than your images. A more classical dilation of 1 make all the timings more consistent and closer what you expect.
- In your timings, note that replicate is 10x faster for 1d so this is not always slower in the replication case. Most likely very bad behavior due to the dilation.

**Comment by helion-du-mas-des-bourboux-thales** [comment]
@albanD, thank you for the comments.
 - I have updated the code with `torch.cuda.synchronize()`
 - The `2**8` dilation is an example to show how bad it can be
 - Indeed the image is smaller that that dilation
 - I still find it strange a so large difference on computation time when changing padding_mode from zeros to replicate

**Comment by ngimel** [comment]
Zero padding is handled implicitly by the libraries implementing convolution (cudnn, OneDnn), replication padding has to be done explicitly before calling the convolution. The relative penalty of this explicit step will heavily depend on actual convolution parameters (e.g. if convolution itself is compute-bound, the padding penalty will be small, if convolution is not as computationally intensive, padding will matter more. In addition, padding itself can be not very efficiently implemented). Since possible space of convolution parameters is extremely large, we don't even attempt to optimize for every case. If you have a realistic range where replication padding performance is problematic, please share.
